test/dist_utils.py

```diff
@@ -47,6 +78,40 @@ def wrapper(self, *arg, **kwargs):
             num_send_recv_threads=16
         )
         test_method(self, *arg, **kwargs)
+
+        # Follower reports done.
+        if self.rank == MASTER_RANK:
+            on_master_follower_report_done(
+                "worker{}".format(MASTER_RANK)
+            )
+        else:
+            rpc.rpc_async(
+                "worker{}".format(MASTER_RANK),
+                on_master_follower_report_done,
+                args=("worker{}".format(self.rank),),
+            )
+
+        # Master waits for followers to report done.
+        # Follower waits for master's termination command.
+        _TERMINATION_SIGNAL.wait()
+        if self.rank == MASTER_RANK:
+            # Master sends termination command.
+            futs = []
+            for dst_rank in range(self.world_size):
+                # torch.distributed.rpc module does not support sending to self.
+                if dst_rank == MASTER_RANK:
+                    continue
+                dst_name = "worker{}".format(dst_rank)
+                fut = rpc.rpc_async(
+                    dst_name,
+                    set_termination_signal,
+                    args=(),
+                )
+                futs.append(fut)
+            for fut in futs:
+                assert fut.wait() is None, "Sending termination signal failed."
+
+        # Close RPC.
         rpc.join_rpc()
```

Comment:
`join_rpc` here still does termination detection, so that we don't know if the above master-based termination would work or not (I think it could still leak futures, no?). Shall we replace this `join_rpc` with a local shutdown?

Reply:
We don't need to delete `join_rpc` implementation in this PR, just add a local shutdown and let's see if it would work.

Reply:
@mrshenli 

That's my plan. 
After 
1) this PR has merged. 
2) local shutdown API is ready.

I will change this to shutdown.


This PR works, because other RpcAgent implement Join as shutdown, and the test_asymmetric_load is failing on the other RpcAgent. 

This PR fix the test, proving it works.

Reply:
I see. That makes sense. I am a little worried if the flakiness in the other RpcAgent is due to leaking futures. Regardless, let's get this in first, then we can fix flakiness, and drop `join_rpc()`.